<a href="https://colab.research.google.com/github/MATHKIMS/DSAC_license/blob/main/%EB%82%98%EC%9D%B4%EB%B8%8C%EB%B2%A0%EC%9D%B4%EC%A6%88%EC%95%8C%EA%B3%A0%EB%A6%AC%EC%A6%98%EC%9D%84_%ED%99%9C%EC%9A%A9%ED%95%9C_%EC%8A%A4%ED%8C%B8%EB%A9%94%EC%9D%BC_%EC%98%88%EC%B8%A1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 베이즈 텍스트 분석

다루는 내용
--

- 메일이 스팸메일인지 아닌지 예측

데이터
--
- https://goo.gl/2WWtcs

학습데이터
양적으로 많아야함

질적으로 정규분포만큼의 질적인 데이터가 존재해야함

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt


In [7]:
from sklearn.model_selection import train_test_split 
from sklearn.feature_extraction.text import CountVectorizer 
from sklearn.naive_bayes import GaussianNB
from sklearn import metrics

ham=정상,spam=스팸

In [12]:
sms=pd.read_csv(r'/content/drive/MyDrive/파이썬/파이썬 기초 (학부)/DSAC 과정/data/sms_spam.csv',encoding='latin-1')
sms.head()

type                                               text
0   ham  Go until jurong point, crazy.. Available only ...
1   ham                      Ok lar... Joking wif u oni...
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...
3   ham  U dun say so early hor... U c already then say...
4   ham  Nah I don't think he goes to usf, he lives aro...

In [14]:
sms.shape #데이터 형태 5574행,2열

(5574, 2)

In [16]:
sms['type'].value_counts() #value_counts메소드 데이터의 개수를 알려줌

ham     4827
spam     747
Name: type, dtype: int64

In [17]:
#데이터를 한꺼번에 뭉침,corpus라 함

sms_corpus= sms['text']

In [18]:
sms_corpus

0       Go until jurong point, crazy.. Available only ...
1                           Ok lar... Joking wif u oni...
2       Free entry in 2 a wkly comp to win FA Cup fina...
3       U dun say so early hor... U c already then say...
4       Nah I don't think he goes to usf, he lives aro...
                              ...                        
5569    This is the 2nd time we have tried 2 contact u...
5570                Will ü b going to esplanade fr home?
5571    Pity, * was in mood for that. So...any other s...
5572    The guy did some bitching but I acted like i'd...
5573                           Rofl. Its true to its name
Name: text, Length: 5574, dtype: object

In [20]:
vectorizer = CountVectorizer(min_df=10) #10번 이하로 발생한 단어 무시

CountVectorizer는 문장을 숫자벡터로 만들어 주는 것임
실제 자연어를 수치배열로 만드는것

fit_transform

fit과 transform 둘다 하는 과정

In [33]:
X=vectorizer.fit_transform(sms_corpus) #
X.shape

(5574, 1020)

5574개의 문장, 1020개의 단어로 만듬

In [34]:
X

<5574x1020 sparse matrix of type '<class 'numpy.int64'>'
	with 58374 stored elements in Compressed Sparse Row format>

In [35]:
y=sms['type']
y.head()

0     ham
1     ham
2    spam
3     ham
4     ham
Name: type, dtype: object

In [36]:
X[:5].toarray() #배열로 변환

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]])

In [37]:
X[:5].toarray().argmax(axis=1) #argmax는 가장 많이 나타낸 단어

array([ 92, 463, 867, 732, 393])

In [40]:
X_train, X_test, y_train, y_test=train_test_split(X.toarray(),y, test_size=0.25) #train_test_split 는 array데이터를 입력받을 수 있다.

In [41]:
clf=GaussianNB()
clf.fit(X_train,y_train)

print(clf.score(X_test,y_test).round(4))

0.8135


GaussianNB모델을 그냥 적용했을때, 81.35%의 성능으로 스팸메일과 정상메일을 분류할 수 있다.

여기서는 텍스트 처리를 전혀하지 않았으나 텍스트 처리를 수행하면
성능을 높일 수 있다!